### DISCUSSION TITLE
Building on Windows using cmake inside VS2019, missing required GLEW lib

### DISCUSSION THREAD

**Discussion by rschader** [projectM-visualizer/projectm#671] [Q&A]
Hello, sorry I am a little rusty on using cmake, especially with Visual Studio. When I open the folder, it goes thru the configuration and halts at an error where it "Could NOT find GLEW (missing: GLEW_INCLUDE_DIRS GLEW_LIBRARIES)". I have downloaded glew and built it using visual studio, but how do I tell cmake where the files are, and does it matter if the lib is w32 or x64? Thanks in advance for your help! 

P.S. I obtained the source just by downloading the code as a zipfile, if that makes any difference. I also thought I would try getting glew with nuget, but nuget didn't like the fact that there is no project file.

**Comment by kblaschke** [comment]
Easiest way in your case is by adding `CMAKE_PREFIX_PATH=<GLEW install dir>` to the CMake configuration parameters. Despite the name, it's a list, so you can also add multiple paths divided by semicolons here, which are searched in order. Useful if you have multiple libraries like GLEW and SDL.

We have detailed build instructions for Windows in the Wiki:

https://github.com/projectM-visualizer/projectm/wiki/Building-libprojectM#building-on-windows

The option described there is installing the dependencies with vcpkg and use the vcpkg toolchain file with your CMake build. I'd recommend this if you're building for different architectures and VC runtimes. Documentation for that is here:

https://vcpkg.io/en/docs/users/buildsystems/cmake-integration.html

**Reply by rschader** [reply]
Thanks! I did manage to get vcpkg installed "and integrated", used that to install glew (looks like it installed it as 32 bit?). Finally got visual studio's cmake to find glew. Still seemed like it would not find it until I found an entry in the cmake config inside visual stugio to also specify the glew install dir, but when it would try to build everything, I would end up with something like 302 warnings and over 100 errors? I haven't looked at your wiki link yet, but I will for sure! I'm thinking it has to be built completey outside of the vs IDE? I haven't done any programming in quite a while, seems like I have a lot of catching up to do!

**Reply by rschader** [reply]
I was able to follow the windows build instructions from the wiki link somewhat successfully. running "vcpkg install glew:x64-windows sdl2:x64-windows" appeared to be successful, but running the cmake command example provided (substituting what I assumed were correct paths), that threw up some errors and told me to look at a couple cmake logfiles. But they seemed to indicate a success? But I could not locate any vs2019 solution or project files? At that point I opened the folder in vs2019, and it successfully went through the cmake config, stating "Build files have been written to: D:/Projects/projectm/out/build/x64-Debug", but again, I could not find any solution or project files to open? However, there was an "Install ProjectM" entry under the build menu, and that seemed to do the trick! Now I just need to look into some example applications that use it, or look at the docs on how to use it in an application. I was kind of expecting that there would have been an example app built into this repository?

**Reply by kblaschke** [reply]
There is an example app, but it's not built by default. To enable it, add `-DENABLE_SDL_UI=YES` to the CMake configuration parameters.

In Visual Studio, you can then select the SDL Test UI as start project and run it. It's kinda broken in current master though, so don't expect it to run too well.

Regarding CMake, vcpkg and Visual Studio, I recommend reading the CMake documentation to get an overview on how it works, it's conceptually a bit different to classic hand-made VS solutions:

https://cmake.org/cmake/help/latest/guide/user-interaction/index.html

**Reply by rschader** [reply]
Thanks! I found an interesting article about projectM last night and it mentioned the SDL test UI app, I went into the github directory for it and see that there is also another repository of a more developed? SDL Frontend that I plan on checking out instead. One last question for no though: The wiki pages for the projectM API reference seems to be blank? Is there anything available out there besides the code? If not, I'll try generating some with doxygen. But I'd rather skip that step if there is something out there?

**Reply by kblaschke** [reply]
Yes, the [SDL2 frontend](https://github.com/projectM-visualizer/frontend-sdl2/) would be the one to check out, as this will be the next-gen user-faced standalone/reference application. It's a full rewrite based on the features of the old UI (which now has become the test UI), but provides way more flexibility to the user, better audio capturing capabilities and cleaner code. Build might be broken against latest projectM master right now, but this will get fixed. Check out [this commit](https://github.com/projectM-visualizer/projectm/commit/62cecb36d6c796771f84511347a5772b48162960) of libprojectM for now, as this is the last commit the SDL2 frontend will build against in the current state.

The API docs aren't there yet as we're still actively in the process of finalizing the API for the next release. Before, projectM only supported a C++ interface with a lot of exposed internals, which made it hard to integrate and impossible to use as a shared library.

That said, if you want to play around with the API, you sure can use Doxygen to generate the docs, as I've added proper comments on all public headers. You can also just read the comments directly in projectM.h, as this is - currently - the only header that contains the API functions. If the API is fixed, we'll add proper documentation in the Wiki with usage examples and further explanations, as some parts like adding audio data isn't always as straightforward as it might seem.

Again, remember the API isn't finished yet and can still change with each commit. After the 4.0 release, we'll keep the API as stable as possible and also try our best to keep it backward compatible.
